src/main/java/org/jenkinsci/plugins/workflow/graphanalysis/ForkScanner.java

```diff
@@ -437,8 +513,71 @@ protected FlowNode next(@Nonnull FlowNode current, @Nonnull Collection<FlowNode>
 
         if (currentParallelStart != null && currentParallelStart.unvisited.size() > 0) {
             output = currentParallelStart.unvisited.pop();
-            currentParallelStart.remainingBranches--;
+            nextType = NodeType.PARALLEL_BRANCH_END;
+        }
+        if (output == null) {
+            nextType = null;
         }
         return output;
     }
+
+    public static void visitSimpleChunks(@Nonnull Collection<FlowNode> heads, @Nonnull Collection<FlowNode> blacklist, @Nonnull SimpleChunkVisitor visitor, @Nonnull ChunkFinder finder) {
+        ForkScanner scanner = new ForkScanner();
+        scanner.setup(heads, blacklist);
+        scanner.visitSimpleChunks(visitor, finder);
+    }
+
+    public static void visitSimpleChunks(@Nonnull Collection<FlowNode> heads, @Nonnull SimpleChunkVisitor visitor, @Nonnull ChunkFinder finder) {
+        ForkScanner scanner = new ForkScanner();
+        scanner.setup(heads);
+        scanner.visitSimpleChunks(visitor, finder);
+    }
+
+    /** Walk through flows  */
+    public void visitSimpleChunks(@Nonnull SimpleChunkVisitor visitor, @Nonnull ChunkFinder finder) {
+        FlowNode prev = null;
+        if (finder.isStartInsideChunk() && hasNext()) {
+            visitor.chunkEnd(this.myNext, null, this);
+        }
+        while(hasNext()) {
+            prev = (myCurrent != myNext) ? myCurrent : null;
+            FlowNode f = next();
+
+            boolean boundary = false;
+            if (finder.isChunkStart(myCurrent, prev)) {
+                visitor.chunkStart(myCurrent, myNext, this);
+                boundary = true;
+            }
+            if (finder.isChunkEnd(myCurrent, prev)) {
+                visitor.chunkEnd(myCurrent, prev, this);
+                boundary = true;
+            }
+            if (!boundary) {
+                visitor.atomNode(myNext, f, prev, this);
+            }
+
+            // Trigger on parallels
+            switch (currentType) {
+                case NORMAL:
+                    break;
+                case PARALLEL_END:
+                    visitor.parallelEnd(this.currentParallelStartNode, myCurrent, this);
+                    break;
+                case PARALLEL_START:
+                    visitor.parallelStart(myCurrent, prev, this);
+                    break;
+                case PARALLEL_BRANCH_END:
+                    visitor.parallelBranchEnd(this.currentParallelStartNode, myCurrent, this);
+                    break;
+                case PARALLEL_BRANCH_START:
+                    // Needed because once we hit the start of the last branch, the next node is our currentParallelStart
+                    FlowNode parallelStart = (nextType == NodeType.PARALLEL_START) ? myNext : this.currentParallelStartNode;
+                    visitor.parallelBranchStart(parallelStart, myCurrent, this);
+                    break;
+                default:
```

Comment:
Good practice: break on NORMAL and throw IllegalStateException on default. Just to avoid missing logic processing if the node types get extended